lib/Common/Memory/Recycler.cpp

```diff
@@ -3072,15 +3079,36 @@ Recycler::Sweep(bool concurrent)
 #if ENABLE_CONCURRENT_GC
     if (concurrent)
     {
-        if (!StartConcurrent(CollectionStateConcurrentSweep))
+        bool needForceForground = false;
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+        if (CONFIG_FLAG_RELEASE(EnableConcurrentSweepAlloc) && this->AllocationsDuringconcurrentSweepTurnedOn())
         {
-           // Failed to spawn the concurrent sweep.
-           // Instead, force the concurrent sweep to happen right here in thread.
-           this->collectionState = CollectionStateConcurrentSweep;
+            needForceForground = !StartConcurrent(CollectionStateConcurrentSweepPass1);
+        }
+        else
+#endif
+        {
+            needForceForground = !StartConcurrent(CollectionStateConcurrentSweep);
+        }
 
-           DoBackgroundWork(true);
-           // Continue as if the concurrent sweep were executing
-           // Next time we check for completion, we will finish the sweep just as if it had happened out of thread.
+        if(needForceForground)
+        {
+            // Failed to spawn the concurrent sweep.
+            // Instead, force the concurrent sweep to happen right here in thread.
+#if ENABLE_ALLOCATIONS_DURING_CONCURRENT_SWEEP
+            if (CONFIG_FLAG_RELEASE(EnableConcurrentSweepAlloc) && this->AllocationsDuringconcurrentSweepTurnedOn())
```

Comment:
What does this mean? We're sweeping here in the foreground because we failed to spawn the background sweep so why do the two-pass sweep here?

Reply:
Good catch. We will just need to make sure that the feature is turned off in this case. I will make this change.

---
In reply to: [148170163](https://github.com/Microsoft/ChakraCore/pull/4104#discussion_r148170163) [](ancestors = 148170163)